File: cmd/server/workitem_query.go
package main

import (
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"math"
	"net/http"
	"sort"
	"strings"
	"time"
)

// Query field names are a closed registry, never interpolated into SQL. Filtering
// and stable sorting run over the complete project-scoped result before paging.
type workItemFilter struct {
	Field    string `json:"field"`
	Operator string `json:"operator"`
	Value    any    `json:"value,omitempty"`
}
type workItemQuery struct {
	Fields      map[string]string
	Filters     []workItemFilter
	Sort, Order string
	Timezone    *time.Location
}
type workQueryValidationError string

func (e workQueryValidationError) Error() string { return string(e) }
func failWorkQuery(w http.ResponseWriter, err error) {
	var validation workQueryValidationError
	if errors.As(err, &validation) {
		fail(w, 422, "invalid_list_query", err.Error())
	} else {
		fail(w, 503, "database_unavailable", "数据暂时无法读取，请稍后重试")
	}
}
func baseWorkQueryFields() map[string]string {
	fields := map[string]string{}
	for _, key := range []string{"code", "title", "type", "category", "sprint", "status", "priority", "discipline", "remarks", "description", "acceptance"} {
		fields[key] = "text"
	}
	for _, key := range []string{"parentId", "progress", "estimatedHours", "actualHours", "weightTotal"} {
		fields[key] = "number"
	}
	for _, key := range []string{"startDate", "endDate", "createdAt", "updatedAt"} {
		fields[key] = "date"
	}
	for _, key := range []string{"sensitive", "authImpact"} {
		fields[key] = "boolean"
	}
	fields["owner"] = "person"
	fields["assignee"] = "person"
	fields["tags"] = "multi"
	for _, role := range requirementWeightRoles {
		fields["role."+role+".value"] = "number"
		fields["role."+role+".userId"] = "person"
	}
	return fields
}
func (a *App) parseRequirementListQuery(r *http.Request) (*workItemQuery, error) {
	query := &workItemQuery{Fields: baseWorkQueryFields(), Sort: r.URL.Query().Get("sort"), Order: r.URL.Query().Get("order")}
	rows, err := a.db.QueryContext(r.Context(), `SELECT key,type FROM field_definitions WHERE tenant_id=? AND project_id=? AND object_type='requirement' AND enabled=1`, tenantID, a.pid())
	if err != nil {
		return nil, err
	}
	for rows.Next() {
		var key, kind string
		if err = rows.Scan(&key, &kind); err != nil {
			rows.Close()
			return nil, err
		}
		switch kind {
		case "number", "date", "boolean":
		case "multi_select", "users":
			kind = "multi"
		default:
			kind = "text"
		}
		query.Fields["cf."+key] = kind
	}
	err = rows.Err()
	rows.Close()
	if err != nil {
		return nil, err
	}
	raw := r.URL.Query().Get("filters")
	if len(raw) > 32768 {
		return nil, workQueryValidationError("筛选条件过长")
	}
	if raw != "" {
		decoder := json.NewDecoder(strings.NewReader(raw))
		decoder.DisallowUnknownFields()
		if decoder.Decode(&query.Filters) != nil {
			return nil, workQueryValidationError("筛选条件格式无效")
		}
		var trailing any
		if decoder.Decode(&trailing) != io.EOF {
			return nil, workQueryValidationError("筛选条件格式无效")
		}
	}
	if len(query.Filters) > 30 {
		return nil, workQueryValidationError("最多 30 个筛选条件")
	}
	for _, filter := range query.Filters {
		if err = validateWorkFilter(filter, query.Fields); err != nil {
			return nil, err
		}
	}
	for _, filter := range query.Filters {
		if query.Fields[filter.Field] == "date" {
			var timezone string
			if err := a.db.QueryRowContext(r.Context(), `SELECT timezone FROM users WHERE tenant_id=? AND id=?`, tenantID, a.uid()).Scan(&timezone); err != nil {
				return nil, err
			}
			query.Timezone, err = time.LoadLocation(timezone)
			if err != nil {
				query.Timezone = time.UTC
			}
			break
		}
	}
	if query.Sort == "" {
		query.Sort = "updatedAt"
	}
	if _, ok := query.Fields[query.Sort]; !ok {
		return nil, workQueryValidationError("排序字段无效或已停用")
	}
	if query.Order == "" {
		query.Order = "desc"
	}
	if query.Order != "asc" && query.Order != "desc" {
		return nil, workQueryValidationError("排序方向须为 asc 或 desc")
	}
	return query, nil
}
func validateWorkFilter(filter workItemFilter, fields map[string]string) error {
	kind, ok := fields[filter.Field]
	if !ok {
		return workQueryValidationError("筛选字段无效或已停用")
	}
	op := filter.Operator
	if op == "is_empty" || op == "not_empty" {
		if filter.Value != nil {
			return workQueryValidationError("空值条件不能携带筛选值")
		}
		return nil
	}
	allowed := map[string]bool{"eq": true, "neq": true}
	switch kind {
	case "number", "date":
		for _, op := range []string{"gt", "gte", "lt", "lte"} {
			allowed[op] = true
		}
	case "person", "multi":
		allowed = map[string]bool{"includes": true, "not_includes": true}
	case "text":
		allowed["contains"] = true
		allowed["not_contains"] = true
	}
	if !allowed[op] {
		return workQueryValidationError("筛选条件不适用于此字段")
	}
	switch kind {
	case "number":
		value, valid := filter.Value.(float64)
		if !valid || math.IsNaN(value) || math.IsInf(value, 0) {
			return workQueryValidationError("数值筛选须使用有效数字")
		}
	case "boolean":
		if _, valid := filter.Value.(bool); !valid {
			return workQueryValidationError("布尔筛选须使用是或否")
		}
	default:
		value, valid := filter.Value.(string)
		if !valid || strings.TrimSpace(value) == "" || len(value) > 4000 {
			return workQueryValidationError("筛选值须为非空文本")
		}
		if kind == "date" {
			if _, err := time.Parse("2006-01-02", value); err != nil {
				return workQueryValidationError("日期筛选须为有效日期")
			}
		}
	}
	return nil
}
func workRecord(value any) (map[string]any, error) {
	raw, err := json.Marshal(value)
	if err != nil {
		return nil, err
	}
	record := map[string]any{}
	err = json.Unmarshal(raw, &record)
	return record, err
}
func workStrings(value any) []string {
	out := []string{}
	if values, ok := value.([]any); ok {
		for _, v := range values {
			if str, ok := v.(string); ok && str != "" {
				out = append(out, str)
			}
		}
	}
	return out
}
func workValue(record map[string]any, key string) any {
	if strings.HasPrefix(key, "cf.") {
		fields, _ := record["customFields"].(map[string]any)
		return fields[strings.TrimPrefix(key, "cf.")]
	}
	if strings.HasPrefix(key, "role.") {
		parts := strings.Split(key, ".")
		weights, _ := record["roleWeights"].(map[string]any)
		role, _ := weights[parts[1]].(map[string]any)
		if parts[2] == "value" {
			return role["value"]
		}
		ids := workStrings(role["userIds"])
		if len(ids) == 0 {
			if id, ok := role["userId"].(string); ok && id != "" {
				ids = append(ids, id)
			}
		}
		return ids
	}
	if key == "owner" || key == "assignee" {
		ids := workStrings(record[key+"UserIds"])
		if len(ids) == 0 {
			if id, ok := record[key+"UserId"].(string); ok && id != "" {
				ids = append(ids, id)
			}
		}
		if len(ids) == 0 {
			if name, ok := record[key].(string); ok && name != "" {
				ids = append(ids, "legacy:"+name)
			}
		}
		return ids
	}
	if key == "tags" {
		tags, _ := record[key].(string)
		parts := strings.FieldsFunc(tags, func(r rune) bool { return r == ',' || r == '，' || r == '\n' })
		out := []string{}
		for _, part := range parts {
			if tag := strings.TrimSpace(part); tag != "" {
				out = append(out, tag)
			}
		}
		return out
	}
	return record[key]
}
func emptyWorkValue(value any) bool {
	if value == nil {
		return true
	}
	switch v := value.(type) {
	case string:
		return v == ""
	case []string:
		return len(v) == 0
	case []any:
		return len(v) == 0
	}
	return false
}
func compareWorkValue(a, b any, kind string) int {
	if kind == "number" {
		x, _ := a.(float64)
		y, _ := b.(float64)
		if x < y {
			return -1
		}
		if x > y {
			return 1
		}
		return 0
	}
	if kind == "boolean" {
		x, _ := a.(bool)
		y, _ := b.(bool)
		if x == y {
			return 0
		}
		if x {
			return 1
		}
		return -1
	}
	text := func(value any) string {
		switch v := value.(type) {
		case []string:
			return strings.Join(v, "\x00")
		case []any:
			parts := []string{}
			for _, p := range v {
				parts = append(parts, fmt.Sprint(p))
			}
			return strings.Join(parts, "\x00")
		case string:
			return strings.ToLower(v)
		default:
			return fmt.Sprint(v)
		}
	}
	return strings.Compare(text(a), text(b))
}
func matchesWorkItem(record map[string]any, filter workItemFilter, kind string, timezone *time.Location) bool {
	value := workValue(record, filter.Field)
	empty := emptyWorkValue(value)
	if filter.Operator == "is_empty" {
		return empty
	}
	if filter.Operator == "not_empty" {
		return !empty
	}
	if empty {
		return false
	}
	if filter.Operator == "includes" || filter.Operator == "not_includes" {
		found := false
		switch values := value.(type) {
		case []string:
			for _, v := range values {
				if v == filter.Value {
					found = true
				}
			}
		case []any:
			for _, v := range values {
				if v == filter.Value {
					found = true
				}
			}
		}
		if filter.Operator == "not_includes" {
			return !found
		}
		return found
	}
	if kind == "date" {
		if text, ok := value.(string); ok && len(text) >= 10 {
			if instant, err := time.Parse(time.RFC3339Nano, text); err == nil && timezone != nil {
				value = instant.In(timezone).Format("2006-01-02")
			} else {
				value = text[:10]
			}
		}
	}
	compared := compareWorkValue(value, filter.Value, kind)
	switch filter.Operator {
	case "eq":
		return compared == 0
	case "neq":
		return compared != 0
	case "gt":
		return compared > 0
	case "gte":
		return compared >= 0
	case "lt":
		return compared < 0
	case "lte":
		return compared <= 0
	case "contains":
		return strings.Contains(strings.ToLower(fmt.Sprint(value)), strings.ToLower(fmt.Sprint(filter.Value)))
	case "not_contains":
		return !strings.Contains(strings.ToLower(fmt.Sprint(value)), strings.ToLower(fmt.Sprint(filter.Value)))
	}
	return false
}
func workSortValue(record map[string]any, key string, names map[string]string) any {
	if key == "code" {
		return record["id"]
	}
	value := workValue(record, key)
	if key == "owner" || key == "assignee" || strings.HasSuffix(key, ".userId") {
		ids, _ := value.([]string)
		out := []string{}
		for _, id := range ids {
			name := names[id]
			if name == "" {
				name = strings.TrimPrefix(id, "legacy:")
			}
			out = append(out, name)
		}
		return out
	}
	return value
}
func (a *App) applyRequirementListQuery(items []Requirement, query *workItemQuery) ([]Requirement, error) {
	type entry struct {
		item   Requirement
		record map[string]any
	}
	entries := []entry{}
	for _, item := range items {
		record, err := requirementQueryRecord(item, query)
		if err != nil {
			return nil, err
		}
		matches := true
		for _, filter := range query.Filters {
			if !matchesWorkItem(record, filter, query.Fields[filter.Field], query.Timezone) {
				matches = false
				break
			}
		}
		if matches {
			entries = append(entries, entry{item, record})
		}
	}
	names := map[string]string{}
	if query.Fields[query.Sort] == "person" {
		rows, err := a.db.Query(`SELECT id,name FROM users WHERE tenant_id=?`, tenantID)
		if err != nil {
			return nil, err
		}
		for rows.Next() {
			var id, name string
			if err = rows.Scan(&id, &name); err != nil {
				rows.Close()
				return nil, err
			}
			names[id] = name
		}
		err = rows.Err()
		rows.Close()
		if err != nil {
			return nil, err
		}
	}
	sort.SliceStable(entries, func(i, j int) bool {
		left, right := entries[i], entries[j]
		x, y := workSortValue(left.record, query.Sort, names), workSortValue(right.record, query.Sort, names)
		if emptyWorkValue(x) != emptyWorkValue(y) {
			return !emptyWorkValue(x)
		}
		kind := query.Fields[query.Sort]
		if query.Sort == "code" {
			kind = "number"
		}
		comparison := compareWorkValue(x, y, kind)
		if comparison == 0 {
			return left.item.ID < right.item.ID
		}
		if query.Order == "desc" {
			return comparison > 0
		}
		return comparison < 0
	})
	result := make([]Requirement, 0, len(entries))
	for _, entry := range entries {
		result = append(result, entry.item)
	}
	return result, nil
}

File: src/workItemQuery.ts
// 列、过滤器和排序共用此白名单。key/value 参与业务比较，label 可本地化，二者不可混用。
// 混合迭代列表的自定义字段带对象类型前缀，防止需求和缺陷同 key 相互覆盖。
export type WorkFieldKind = 'text' | 'number' | 'date' | 'boolean' | 'person' | 'multi'
export type WorkFilter = { field: string; operator: string; value?: string | number | boolean }
export type WorkField = { key: string; label: string; kind: WorkFieldKind; group?: string; options?: { value: string; label: string }[]; systemOptions?: boolean; custom?: boolean; width?: number; default?: boolean; fixed?: boolean }
export const workRoleFields = [
  ['frontend','前端开发难度','前端工程师'], ['backend','后端难度','后端工程师'], ['ui','UI 难度','UI 工程师'], ['algorithm','算法难度','算法工程师'], ['product','产品难度','产品负责人'],
] as const
const core: WorkField[] = [
  { key:'code',label:'编号',kind:'text',fixed:true,width:115 }, { key:'title',label:'标题',kind:'text',fixed:true,width:300 },
  { key:'objectType',label:'工作项类型',kind:'text',default:true,options:[{value:'requirement',label:'需求'},{value:'defect',label:'缺陷'}],systemOptions:true },
  { key:'type',label:'需求类型',kind:'text' }, { key:'category',label:'分类',kind:'text' },
  { key:'status',label:'状态',kind:'text',default:true }, { key:'priority',label:'优先级',kind:'text',default:true,options:['P0','P1','P2','P3'].map(value=>({value,label:value})) },
  { key:'sprint',label:'迭代',kind:'text' }, { key:'owner',label:'产品负责人',kind:'person' }, { key:'assignee',label:'处理人',kind:'person',default:true,width:190 },
  { key:'discipline',label:'职能',kind:'text',options:Object.entries({product:'产品',frontend:'前端',backend:'后端',algorithm:'算法',ui:'UI',qa:'测试'}).map(([value,label])=>({value,label})),systemOptions:true },
  ...workRoleFields.flatMap(([key,label,person]):WorkField[]=>[{key:`role.${key}.userId`,label:person,kind:'person',group:'角色权重',width:170},{key:`role.${key}.value`,label,kind:'number',group:'角色权重',default:true,width:130}]),
  { key:'weightTotal',label:'总权重',kind:'number',default:true,group:'角色权重' },
  { key:'tags',label:'标签',kind:'multi',width:180 }, { key:'remarks',label:'备注',kind:'text',width:230 },
  { key:'description',label:'需求描述',kind:'text',width:260 }, { key:'acceptance',label:'验收标准',kind:'text',width:230 },
  { key:'parentId',label:'父需求',kind:'number' }, { key:'progress',label:'进度',kind:'number' },
  { key:'estimatedHours',label:'预估工时',kind:'number' }, { key:'actualHours',label:'实际工时',kind:'number' },
  { key:'sensitive',label:'涉及敏感数据',kind:'boolean' }, { key:'authImpact',label:'涉及权限认证',kind:'boolean' },
  { key:'startDate',label:'计划开始',kind:'date' }, { key:'endDate',label:'计划结束',kind:'date' },
  { key:'createdAt',label:'创建时间',kind:'date',width:180 }, { key:'updatedAt',label:'更新时间',kind:'date',width:180 },
]
const defectFields: WorkField[] = [
  {key:'severity',label:'严重程度',kind:'text'}, {key:'verifier',label:'验证人',kind:'text'}, {key:'requirementId',label:'关联需求',kind:'number'},
  {key:'steps',label:'复现步骤',kind:'text'}, {key:'actual',label:'实际结果',kind:'text'}, {key:'expected',label:'预期结果',kind:'text'},
  {key:'environment',label:'环境',kind:'text'}, {key:'foundVersion',label:'发现版本',kind:'text'}, {key:'fixVersion',label:'修复版本',kind:'text'},
]
export function workItemFields(definitions:any[]=[], members:any[]=[], mixed=false):WorkField[] {
  // 人员候选以稳定用户 ID 为值，部门约束来自真实 departmentIds；姓名只显示。
  // 可选候选不等于授权：保存时服务端仍应校验活动成员、项目和字段部门约束。
  const people=members.map(member=>({value:member.id,label:member.name+' · '+(member.department||member.email||member.id)}))
  // 迭代默认顺序：编号、标题、创建时间；不覆盖用户已经保存的列顺序。
  // 恢复默认与首次加载共用此字段表，创建时间仍可隐藏、移动。
  const base=mixed?[...core.slice(0,2),{...core.find(field=>field.key==='createdAt')!,default:true},...core.slice(2).filter(field=>field.key!=='createdAt')]:core.filter(field=>field.key!=='objectType')
  return [...base.map(field=>({...field,group:field.group||'基础信息',options:field.kind==='person'?people:field.options})),...(mixed?defectFields.map(field=>({...field,group:'缺陷字段'})):[]),
    ...definitions.filter(definition=>definition.enabled!==false).map(definition=>({
      key:'cf.'+(mixed?definition.objectType+'.':'')+definition.key,label:definition.name,kind:({number:'number',date:'date',boolean:'boolean',multi_select:'multi',users:'multi'} as Record<string,WorkFieldKind>)[definition.type]||'text',
      options:['user','users'].includes(definition.type)?members.filter(member=>member.active!==false&&(!definition.departmentId||member.departmentIds?.includes(definition.departmentId))).map(member=>({value:member.id,label:member.name+' · '+(member.departmentNames?.join('、')||member.email||member.id)})):(definition.options||[]).map((value:string)=>({value,label:value})),
      group:mixed?(definition.objectType==='defect'?'缺陷自定义字段':'需求自定义字段'):'自定义字段',custom:true,default:definition.listVisible,width:160,
    }))]
}
export const operatorLabels:Record<string,string>={eq:'等于',neq:'不等于',contains:'包含文本',not_contains:'不包含文本',gt:'大于',gte:'大于或等于',lt:'小于',lte:'小于或等于',includes:'包含成员或选项',not_includes:'不包含成员或选项',is_empty:'为空',not_empty:'不为空'}
export function workOperators(field:WorkField):string[]{return [...(field.kind==='number'||field.kind==='date'?['eq','neq','gt','gte','lt','lte']:field.kind==='boolean'?['eq','neq']:['person','multi'].includes(field.kind)?['includes','not_includes']:field.options?.length?['eq','neq']:['contains','not_contains','eq','neq']),'is_empty','not_empty']}
export function workFilterValue(field:WorkField,operator:string,value:unknown):WorkFilter['value']{
  // 空值判断操作不携带 value；数值 0、布尔 false 都是有效条件，不能用 truthy 判断。
  if(['is_empty','not_empty'].includes(operator))return undefined
  if(field.kind==='number'){if((typeof value==='string'&&!value.trim())||value==null||!Number.isFinite(Number(value)))throw new Error('请输入有效数值');return Number(value)}
  if(field.kind==='boolean'){if(value!==true&&value!==false&&value!=='true'&&value!=='false')throw new Error('请选择是或否');return value===true||value==='true'}
  if(typeof value!=='string'||!value.trim())throw new Error('请输入筛选值')
  if(field.kind==='date'&&(!/^\d{4}-\d{2}-\d{2}$/.test(value)||!Number.isFinite(Date.parse(value))||new Date(value).toISOString().slice(0,10)!==value))throw new Error('请选择有效日期')
  return value
}
function personIDs(item:any,kind:string):string[]{const values=item[kind+'UserIds'];return Array.isArray(values)&&values.length?[...new Set(values)]:(item[kind+'UserId']?[item[kind+'UserId']]:item[kind]?['legacy:'+item[kind]]:[])}
// 兼容多人数组、旧单 ID 和历史姓名；legacy: 仅供显示/比较，不能当成真实用户 ID 提交。
export function workItemValue(item:any,key:string):any {
  if(key.startsWith('cf.')){const parts=key.split('.');return parts.length===3?(parts[1]===item.objectType?item.customFields?.[parts[2]!]:null):item.customFields?.[key.slice(3)]}
  if(key.startsWith('role.')){const [,role,property]=key.split('.'),weight=item.roleWeights?.[role!];return property==='userId'?(weight?.userIds?.length?weight.userIds:weight?.userId?[weight.userId]:[]):weight?.value}
  if(key==='assignee'||key==='owner')return personIDs(item,key)
  if(key==='tags')return String(item.tags||'').split(/[,，\n]/).map(value=>value.trim()).filter(Boolean)
  return item[key]
}
export function workItemPersonNames(item:any,key:string,members:any[]=[]):string[]{
  const ids=workItemValue(item,key),snapshots=key==='assignee'?item.assignees:key==='owner'?item.owners:[]
  return Array.isArray(ids)&&ids.length?ids.map(id=>members.find(member=>member.id===id)?.name||snapshots?.find((person:any)=>person.id===id)?.name||id.replace(/^legacy:/,'')):(key==='assignee'||key==='owner')&&item[key]?[item[key]]:[]
}
export function emptyWorkValue(value:any):boolean{return value==null||value===''||(Array.isArray(value)&&!value.length)}
function scalar(value:any,kind:WorkFieldKind){return kind==='date'?String(value).slice(0,10):kind==='text'?String(value).toLowerCase():Array.isArray(value)?value.join('\u0000'):value}
export function workCalendarDate(value:any,timezone='UTC'):string{
  // 计划日期 YYYY-MM-DD 不做时区平移；创建时间等时间戳按账号时区归属自然日。
  const text=String(value);if(/^\d{4}-\d{2}-\d{2}$/.test(text))return text
  const date=new Date(text);if(!Number.isFinite(date.getTime()))return text.slice(0,10)
  const parts=new Intl.DateTimeFormat('en-US',{timeZone:timezone,year:'numeric',month:'2-digit',day:'2-digit'}).formatToParts(date)
  return ['year','month','day'].map(type=>parts.find(part=>part.type===type)?.value).join('-')
}
export function matchesWorkFilter(item:any,rule:WorkFilter,field:WorkField,timezone='UTC'):boolean{
  const value=workItemValue(item,rule.field),empty=emptyWorkValue(value)
  if(rule.operator==='is_empty')return empty
  if(rule.operator==='not_empty')return !empty
  if(empty)return false
  if(rule.operator==='includes'||rule.operator==='not_includes'){const has=Array.isArray(value)&&value.includes(rule.value);return rule.operator==='includes'?has:!has}
  const actual=field.kind==='date'?workCalendarDate(value,timezone):scalar(value,field.kind),expected=scalar(rule.value,field.kind)
  switch(rule.operator){case'eq':return actual===expected;case'neq':return actual!==expected;case'contains':return String(actual).includes(String(expected));case'not_contains':return !String(actual).includes(String(expected));case'gt':return actual>expected;case'gte':return actual>=expected;case'lt':return actual<expected;case'lte':return actual<=expected;default:return false}
}
export function queryWorkItems(items:any[],fields:WorkField[],rules:WorkFilter[],sortKey:string,order:string,members:any[]=[],timezone='UTC'):any[]{
  // 多个字段条件为 AND；未知字段直接不匹配，避免字段删除后静默扩大结果。
  // 这是当前已授权数据的本地查询；服务端分页列表需服务端完成同口径过滤后再分页。
  const registry=new Map(fields.map(field=>[field.key,field])),sortField=registry.get(sortKey)
  const result=items.filter(item=>rules.every(rule=>{const field=registry.get(rule.field);return !!field&&matchesWorkFilter(item,rule,field,timezone)}))
  return result.sort((left,right)=>{
    const a=sortKey==='code'?left.id:sortField?.kind==='person'?workItemPersonNames(left,sortKey,members):workItemValue(left,sortKey),b=sortKey==='code'?right.id:sortField?.kind==='person'?workItemPersonNames(right,sortKey,members):workItemValue(right,sortKey)
    if(emptyWorkValue(a)!==emptyWorkValue(b))return emptyWorkValue(a)?1:-1
    let compared=0
    if(!emptyWorkValue(a)){const kind=sortKey==='code'?'number':sortField?.kind||'text',x=kind==='date'?String(a):scalar(a,kind),y=kind==='date'?String(b):scalar(b,kind);compared=x<y?-1:x>y?1:0}
    return compared*(order==='asc'?1:-1)||String(left.objectType||'').localeCompare(String(right.objectType||''))||left.id-right.id
  })
}
